### ISSUE TITLE
Libvirt e2e with kcli no longer works since CoCo 0.8.0

### ISSUE THREAD

**Issue by wainersm** [confidential-containers/cloud-api-adaptor#1593]
As of CoCo 0.8.0 it is required containerd 1.7+.

The Libivrt implementation of our e2e test framework relies on kcli to deploy Kubernetes on VMs. Currently worker nodes are installed with Ubuntu 20.04 and containerd 1.6. Ubuntu 22.04  provides containerd 1.7 (https://packages.ubuntu.com/jammy-updates/containerd), so one solution to meet the containerd requirement would be to simple update Ubuntu version, but....

.... kcli has installed containerd.io package which is distributed by Docker, which is containerd 1.6. I opened an issue on kcli side (https://github.com/karmab/kcli/issues/619) to see if we can work around that situation, perhaps allowing users to choose the containerd package/version to be installed on the nodes.

An alternative solution would be to pass the `INSTALL_OFFICIAL_CONTAINERD=true` to CoCo's operator so that it will swap the system's containerd.  This is a band-aid solution though, we don't know for how long the operator will keep shipping with that containerd switch feature.

Another alternative solution is to look for another tool to deploy Kubernetes.

**Comment by huoqifeng** [comment]
Thanks @wainersm , CC @liudalibj @stevenhorsman

**Comment by stevenhorsman** [comment]
@wainersm - Hey Wainer, as you know I've been playing around with running the libvirt e2e tests on s390x and kcli doesn't support this, I've used the same approach as we use in the operator to set-up the cluster with ansible and kubeadm. When running this on a Ubuntu 22.04 machine this all works. My history is in this gist: https://gist.github.com/stevenhorsman/c894345ef7737585ffaf761f74bd80d7 in case that could be an approach to use on all platforms?

Alternatively we could consider skipping the nydus test on the libvirt cloud provider until we have https://github.com/confidential-containers/operator/issues/288 implemented, or a better solution?

**Comment by karmab** [comment]
I believe kcli should work on s390x too!

**Comment by stevenhorsman** [comment]
@karmab - Hey Karim, that's interesting to know. The issue I hit during the install was:
```
curl https://raw.githubusercontent.com/karmab/kcli/main/install.sh | sudo bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4694  100  4694    0     0  22246      0 --:--:-- --:--:-- --:--:-- 22141
Installing using deb package
Executing the  setup script for the 'karmab/kcli' repository ...

   OK: Checking for required executable 'curl' ...
   OK: Checking for required executable 'apt-get' ...
   OK: Detecting your OS distribution and release using system methods ...
 ^^^^: ... Detected/provided for your OS/distribution, version and architecture:
 >>>>:
 >>>>: ... distro=ubuntu  version=20.04  codename=focal  arch=s390x
 >>>>:
 NOPE: Checking for apt dependency 'apt-transport-https' ...
   OK: Updating apt repository metadata cache ...
   OK: Attempting to install 'apt-transport-https' ...
   OK: Checking for apt dependency 'ca-certificates' ...
   OK: Checking for apt dependency 'gnupg' ...
   OK: Checking for apt signed-by key support ...
   OK: Importing 'karmab/kcli' repository GPG keys ...
   OK: Checking if upstream install config is OK ...
   OK: Installing 'karmab/kcli' repository via apt ...
 NOPE: Updating apt repository metadata cache ...
Failed to update via apt-get update
Hit:1 https://download.docker.com/linux/ubuntu focal InRelease
Hit:2 http://gb.ports.ubuntu.com/ubuntu-ports focal InRelease
Hit:3 http://gb.ports.ubuntu.com/ubuntu-ports focal-updates InRelease
Hit:4 http://gb.ports.ubuntu.com/ubuntu-ports focal-backports InRelease
Hit:5 http://gb.ports.ubuntu.com/ubuntu-ports focal-security InRelease
Get:6 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease [4972 B]
Err:6 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
Reading package lists...
W: GPG error: https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
E: The repository 'https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease' is not signed.
 ^^^^: ... Failed to update via apt-get update - Context above.
 !!!!: Oh no, your setup failed! :-( ... But we might be able to help. :-)
 !!!!:
 !!!!: You can contact karmab for further assistance.
 !!!!:
 !!!!: URL: https://github.com/karmab/kcli
 !!!!:

Get:1 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease [4972 B]
Err:1 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
Hit:2 https://download.docker.com/linux/ubuntu focal InRelease
Hit:3 http://gb.ports.ubuntu.com/ubuntu-ports focal InRelease
Hit:4 http://gb.ports.ubuntu.com/ubuntu-ports focal-updates InRelease
Hit:5 http://gb.ports.ubuntu.com/ubuntu-ports focal-backports InRelease
Hit:6 http://gb.ports.ubuntu.com/ubuntu-ports focal-security InRelease
Reading package lists... Done
W: GPG error: https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
E: The repository 'https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package python3-kcli
```

I assumed it just wasn't supposed to work, but I can raise an issue for this if you'd like?

**Comment by liudalibj** [comment]
@karmab @stevenhorsman I tried the kcli in past days, and figure how to use kcli to create kubeadm s390x cluster on one s390x machine. I created on pr https://github.com/karmab/kcli/pull/620 to fix the issues that I founded. Please help to review and comments.

**Comment by karmab** [comment]
> @karmab - Hey Karim, that's interesting to know. The issue I hit during the install was:
> 
> ```
> curl https://raw.githubusercontent.com/karmab/kcli/main/install.sh | sudo bash
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100  4694  100  4694    0     0  22246      0 --:--:-- --:--:-- --:--:-- 22141
> Installing using deb package
> Executing the  setup script for the 'karmab/kcli' repository ...
> 
>    OK: Checking for required executable 'curl' ...
>    OK: Checking for required executable 'apt-get' ...
>    OK: Detecting your OS distribution and release using system methods ...
>  ^^^^: ... Detected/provided for your OS/distribution, version and architecture:
>  >>>>:
>  >>>>: ... distro=ubuntu  version=20.04  codename=focal  arch=s390x
>  >>>>:
>  NOPE: Checking for apt dependency 'apt-transport-https' ...
>    OK: Updating apt repository metadata cache ...
>    OK: Attempting to install 'apt-transport-https' ...
>    OK: Checking for apt dependency 'ca-certificates' ...
>    OK: Checking for apt dependency 'gnupg' ...
>    OK: Checking for apt signed-by key support ...
>    OK: Importing 'karmab/kcli' repository GPG keys ...
>    OK: Checking if upstream install config is OK ...
>    OK: Installing 'karmab/kcli' repository via apt ...
>  NOPE: Updating apt repository metadata cache ...
> Failed to update via apt-get update
> Hit:1 https://download.docker.com/linux/ubuntu focal InRelease
> Hit:2 http://gb.ports.ubuntu.com/ubuntu-ports focal InRelease
> Hit:3 http://gb.ports.ubuntu.com/ubuntu-ports focal-updates InRelease
> Hit:4 http://gb.ports.ubuntu.com/ubuntu-ports focal-backports InRelease
> Hit:5 http://gb.ports.ubuntu.com/ubuntu-ports focal-security InRelease
> Get:6 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease [4972 B]
> Err:6 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease
>   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> Reading package lists...
> W: GPG error: https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> E: The repository 'https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease' is not signed.
>  ^^^^: ... Failed to update via apt-get update - Context above.
>  !!!!: Oh no, your setup failed! :-( ... But we might be able to help. :-)
>  !!!!:
>  !!!!: You can contact karmab for further assistance.
>  !!!!:
>  !!!!: URL: https://github.com/karmab/kcli
>  !!!!:
> 
> Get:1 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease [4972 B]
> Err:1 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease
>   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> Hit:2 https://download.docker.com/linux/ubuntu focal InRelease
> Hit:3 http://gb.ports.ubuntu.com/ubuntu-ports focal InRelease
> Hit:4 http://gb.ports.ubuntu.com/ubuntu-ports focal-updates InRelease
> Hit:5 http://gb.ports.ubuntu.com/ubuntu-ports focal-backports InRelease
> Hit:6 http://gb.ports.ubuntu.com/ubuntu-ports focal-security InRelease
> Reading package lists... Done
> W: GPG error: https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> E: The repository 'https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease' is not signed.
> N: Updating from such a repository can't be done securely, and is therefore disabled by default.
> N: See apt-secure(8) manpage for repository creation and user configuration details.
> Reading package lists... Done
> Building dependency tree
> Reading state information... Done
> E: Unable to locate package python3-kcli
> ```
> 
> I assumed it just wasn't supposed to work, but I can raise an issue for this if you'd like?

I dont know why that part doesnt work on s390
but the following can be used as a workaround (so you install the deb package and not from pip)

```
curl -1sLf 'curl -sL 'https://dl.cloudsmith.io/public/karmab/kcli/gpg.F9EA2C192F1D6BB6.key'' | sudo gpg --dearmor -o /usr/share/keyrings/kcli-stable-archive-keyring.gpg
```

**Comment by liudalibj** [comment]
> @karmab @stevenhorsman I tried the kcli in past days, and figure how to use kcli to create kubeadm s390x cluster on one s390x machine. I created one pr [karmab/kcli#620](https://github.com/karmab/kcli/pull/620) to fix the issues that I founded. Please help to review and comments.

@karmab thanks for review and merge the s390x support pr, the code introduced one bug: the amd64 KVM vm instance will have bad nicname `enp1s0`, the correct name should be `ens3`, a fix is here https://github.com/karmab/kcli/pull/622

**Comment by liudalibj** [comment]
> > @karmab - Hey Karim, that's interesting to know. The issue I hit during the install was:
> > ```
> > curl https://raw.githubusercontent.com/karmab/kcli/main/install.sh | sudo bash
> >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
> >                                  Dload  Upload   Total   Spent    Left  Speed
> > 100  4694  100  4694    0     0  22246      0 --:--:-- --:--:-- --:--:-- 22141
> > Installing using deb package
> > Executing the  setup script for the 'karmab/kcli' repository ...
> > 
> >    OK: Checking for required executable 'curl' ...
> >    OK: Checking for required executable 'apt-get' ...
> >    OK: Detecting your OS distribution and release using system methods ...
> >  ^^^^: ... Detected/provided for your OS/distribution, version and architecture:
> >  >>>>:
> >  >>>>: ... distro=ubuntu  version=20.04  codename=focal  arch=s390x
> >  >>>>:
> >  NOPE: Checking for apt dependency 'apt-transport-https' ...
> >    OK: Updating apt repository metadata cache ...
> >    OK: Attempting to install 'apt-transport-https' ...
> >    OK: Checking for apt dependency 'ca-certificates' ...
> >    OK: Checking for apt dependency 'gnupg' ...
> >    OK: Checking for apt signed-by key support ...
> >    OK: Importing 'karmab/kcli' repository GPG keys ...
> >    OK: Checking if upstream install config is OK ...
> >    OK: Installing 'karmab/kcli' repository via apt ...
> >  NOPE: Updating apt repository metadata cache ...
> > Failed to update via apt-get update
> > Hit:1 https://download.docker.com/linux/ubuntu focal InRelease
> > Hit:2 http://gb.ports.ubuntu.com/ubuntu-ports focal InRelease
> > Hit:3 http://gb.ports.ubuntu.com/ubuntu-ports focal-updates InRelease
> > Hit:4 http://gb.ports.ubuntu.com/ubuntu-ports focal-backports InRelease
> > Hit:5 http://gb.ports.ubuntu.com/ubuntu-ports focal-security InRelease
> > Get:6 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease [4972 B]
> > Err:6 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease
> >   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> > Reading package lists...
> > W: GPG error: https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> > E: The repository 'https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease' is not signed.
> >  ^^^^: ... Failed to update via apt-get update - Context above.
> >  !!!!: Oh no, your setup failed! :-( ... But we might be able to help. :-)
> >  !!!!:
> >  !!!!: You can contact karmab for further assistance.
> >  !!!!:
> >  !!!!: URL: https://github.com/karmab/kcli
> >  !!!!:
> > 
> > Get:1 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease [4972 B]
> > Err:1 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease
> >   The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> > Hit:2 https://download.docker.com/linux/ubuntu focal InRelease
> > Hit:3 http://gb.ports.ubuntu.com/ubuntu-ports focal InRelease
> > Hit:4 http://gb.ports.ubuntu.com/ubuntu-ports focal-updates InRelease
> > Hit:5 http://gb.ports.ubuntu.com/ubuntu-ports focal-backports InRelease
> > Hit:6 http://gb.ports.ubuntu.com/ubuntu-ports focal-security InRelease
> > Reading package lists... Done
> > W: GPG error: https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F9EA2C192F1D6BB6
> > E: The repository 'https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease' is not signed.
> > N: Updating from such a repository can't be done securely, and is therefore disabled by default.
> > N: See apt-secure(8) manpage for repository creation and user configuration details.
> > Reading package lists... Done
> > Building dependency tree
> > Reading state information... Done
> > E: Unable to locate package python3-kcli
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I assumed it just wasn't supposed to work, but I can raise an issue for this if you'd like?
> 
> I dont know why that part doesnt work on s390 but the following can be used as a workaround (so you install the deb package and not from pip)
> 
> ```
> curl -1sLf 'curl -sL 'https://dl.cloudsmith.io/public/karmab/kcli/gpg.F9EA2C192F1D6BB6.key'' | sudo gpg --dearmor -o /usr/share/keyrings/kcli-stable-archive-keyring.gpg
> ```

- I tried with follow modification to pass the gpg key issue, but still can't install kcli with `apt-get`
```
curl -fsSL https://dl.cloudsmith.io/public/karmab/kcli/gpg.F9EA2C192F1D6BB6.key | sudo gpg --dearmor -o /usr/share/keyrings/kcli-stable-archive-keyring.gpg && apt-key add /usr/share/keyrings/kcli-stable-archive-keyring.gpg
```
- check the key
```
apt-key list
/etc/apt/trusted.gpg
--------------------
pub   rsa2048 2020-11-15 [SCEA]
      91E0 9D37 6184 2688 0765  4901 F9EA 2C19 2F1D 6BB6
uid           [ unknown] Cloudsmith Package (karmab/kcli) <support@cloudsmith.io>

/etc/apt/trusted.gpg.d/ubuntu-keyring-2012-archive.gpg
------------------------------------------------------
pub   rsa4096 2012-05-11 [SC]
      790B C727 7767 219C 42C8  6F93 3B4F E6AC C0B2 1F32
uid           [ unknown] Ubuntu Archive Automatic Signing Key (2012) <ftpmaster@ubuntu.com>

/etc/apt/trusted.gpg.d/ubuntu-keyring-2012-cdimage.gpg
------------------------------------------------------
pub   rsa4096 2012-05-11 [SC]
      8439 38DF 228D 22F7 B374  2BC0 D94A A3F0 EFE2 1092
uid           [ unknown] Ubuntu CD Image Automatic Signing Key (2012) <cdimage@ubuntu.com>

/etc/apt/trusted.gpg.d/ubuntu-keyring-2018-archive.gpg
------------------------------------------------------
pub   rsa4096 2018-09-17 [SC]
      F6EC B376 2474 EDA9 D21B  7022 8719 20D1 991B C93C
uid           [ unknown] Ubuntu Archive Automatic Signing Key (2018) <ftpmaster@ubuntu.com>

```
- update the apt-repo as:
```
cat /etc/apt/sources.list.d/karmab-kcli.list
# Source: karmab
# Site: https://github.com/karmab/kcli
# Repository: karmab / kcli
# Description: A certifiably-awesome open-source package repository curated by karmab, hosted by Cloudsmith.

deb https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal main

deb-src https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal main
```
- run update command
```
apt-get update
Hit:1 http://gb.ports.ubuntu.com/ubuntu-ports focal InRelease
Hit:2 http://gb.ports.ubuntu.com/ubuntu-ports focal-updates InRelease
Get:3 https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease [4972 B]
Hit:4 http://gb.ports.ubuntu.com/ubuntu-ports focal-backports InRelease
Hit:5 http://gb.ports.ubuntu.com/ubuntu-ports focal-security InRelease
Fetched 4972 B in 1s (4010 B/s)
Reading package lists... Done
N: Skipping acquire of configured file 'main/binary-s390x/Packages' as repository 'https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu focal InRelease' doesn't support architecture 's390x'
```
- check the website https://dl.cloudsmith.io/public/karmab/kcli/deb/ubuntu/dists/focal/InRelease
```
...
Architectures: amd64 arm64 armel armhf i386
...
```
So the s390x deb package is not there, is there any way to add s390x deb package @karmab ?
